### PR TITLE
Document describing the process of implementing language features

### PR DESCRIPTION
**Pull request by jaredpar**

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler

File: docs/contributing/Developing a Language Feature.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Developing a Language Feature 
+
+Adding a new feature to C# or VB is a very serious undertaking that often takes several iterations to complete for even the (seemingly) simplest of features. This is due to both the inherent complexity of changing languages and the need to consider the effects of new features in all layers of the Roslyn codebase: IDE, debugging, scripting, etc. As such, language work occurs in a separate branch until the feature reaches a point when we are ready to merge it into the main compiler.
+
+This page discusses the process by which langauge featuere *implementations* are considered, prototyped, and fully accepted into the language.  This process is intended to be used by the compiler team and community.
```
Comment (amcasey) [inline docs/contributing/Developing a Language Feature.md]: "language feature"

File: docs/contributing/Developing a Language Feature.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+Adding a new feature to C# or VB is a very serious undertaking that often takes several iterations to complete for even the (seemingly) simplest of features. This is due to both the inherent complexity of changing languages and the need to consider the effects of new features in all layers of the Roslyn codebase: IDE, debugging, scripting, etc. As such, language work occurs in a separate branch until the feature reaches a point when we are ready to merge it into the main compiler.
+
+This page discusses the process by which langauge featuere *implementations* are considered, prototyped, and fully accepted into the language.  This process is intended to be used by the compiler team and community.  
+
+## Process 
+
+1. **Feature specification filed**:  The speclet should be filed as a GitHub issue and contain:
+
+    a. A description of the feature (including any syntax changes involved)
```
Comment (amcasey) [inline docs/contributing/Developing a Language Feature.md]: Not sure this sublist markup worked.  I'm seeing it as text.

File: docs/contributing/Developing a Language Feature.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    a. Discussions about impacted areas, such as overload resolution and type inference.
+    a. Proposed changes to the API surface area.
+
+    A feature speclet is different from a language design discussion. Discussions are very open-ended and often for features that simply won't work in the language. A feature speclet is a declaration of intent to do the feature as proposed. In many cases, this will come out of a language design discussion. 
+
+    Linking to a early prototype of the proposed feature is highly encouraged. It gives the Language Team a clear understanding of the design intent of the feature and can often serve as the initial commit in the feature branch.
```
Comment (amcasey) [inline docs/contributing/Developing a Language Feature.md]: "an early"

File: docs/contributing/Developing a Language Feature.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+## Frequently Asked Questions
+
+- **Can feature branches be rebased?**: This is left to the discretion of the developer working on the feature. Once the feature is merged into a main branch (master, future) then rebasing is no longer allowed. 
+
+- **Why was my language feature PR closed?**: Language features will not be accepted as direct PRs to a main branch. Even simple features are too involved for senior Language Team members to complete in this fashion. Features need to follow the above process in order to be incorporated into the language.
+
+- **Why are features disabled by default?**: Language features spend a considerable amount of time in the same branch as our shipping product which is often master.  The master branch of Roslyn is kept it a ship ready state at all times.  To facilitate this all features which haven't been fully accepted into the language must be disabled.
```
Comment (BillWagner) [inline docs/contributing/Developing a Language Feature.md]: "it a ship ready state" should be "in a ship ready state".

File: docs/contributing/Developing a Language Feature.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+Adding a new feature to C# or VB is a very serious undertaking that often takes several iterations to complete for even the (seemingly) simplest of features. This is due to both the inherent complexity of changing languages and the need to consider the effects of new features in all layers of the Roslyn codebase: IDE, debugging, scripting, etc. As such, language work occurs in a separate branch until the feature reaches a point when we are ready to merge it into the main compiler.
+
+This page discusses the process by which langauge featuere *implementations* are considered, prototyped, and fully accepted into the language.  This process is intended to be used by the compiler team and community.  
+
+## Process 
+
+1. **Feature specification filed**:  The speclet should be filed as a GitHub issue and contain:
+
+    a. A description of the feature (including any syntax changes involved)
+    a. Discussions about impacted areas, such as overload resolution and type inference.
```
Comment (BillWagner) [inline docs/contributing/Developing a Language Feature.md]: I don't think there can be an exhaustive list of "impacted areas", but I think a second sentence about the complexity of impacted areas for new features is warranted. Maybe something along the lines of "Think through the major areas of the language specification while determine the potential impacts."

File: docs/contributing/Developing a Language Feature.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    
+1. **Prototype phase**: All features are initially prototyped in a separate branch. The name of the branch is `features/<feature name>`. These branches are created when a feature speclet is marked as 'Ready for prototype', language team member champions the feature, and [a developer is assigned](#community-contributions) to implement the feature. The developer is responsible for keeping an up-to-date speclet of the feature at `docs/features/<feature name>.md` throughout the development.  
+ 
+     All features should be off by default until fully accepted into the language. A new feature should only be enabled when explicitly included in the `/features:<feature-name>` compiler option. This helps ensure the language is always in a shippable state no matter what prototypes are in the current branch. 
+
+1. **Prototype decision**: When the prototype phase ends, the Language Team will make a decision on the feature: 
+    * **Ready to integrate**: The feature implementation revealed no blocking issues, has a healthy set of test suites, and is sufficiently complete in the compiler, IDE, debugger, etc. to integrate into the main tree on a selected branch.
```
Comment (BillWagner) [inline docs/contributing/Developing a Language Feature.md]: Should you add something about test suites showing the impact (or proving lack of impact) on other language features?

File: docs/contributing/Developing a Language Feature.md (hunk 2 above)
Comment (jaredpar) [inline docs/contributing/Developing a Language Feature.md]: Looks to be the case.  Rendered in the editor I was using to test but not github.  Trying a new method.

File: docs/contributing/Developing a Language Feature.md (hunk 5 above)
Comment (jaredpar) [inline docs/contributing/Developing a Language Feature.md]: great suggestion

File: docs/contributing/Developing a Language Feature.md (hunk 6 above)
Comment (jaredpar) [inline docs/contributing/Developing a Language Feature.md]: Will add some clarity here.

File: docs/contributing/Developing a Language Feature.md (hunk 4 above)
Comment (jaredpar) [inline docs/contributing/Developing a Language Feature.md]: fixed

**Comment by gafter** [comment]
LGTM

File: docs/contributing/Developing a Language Feature.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    Features which are in the main tree but not tied to a langversion switch are referred to as experimental. They can be pulled or changed at any time.
+
+1. **Acceptance**: When the Language Team feels a feature is complete it will be accepted into the language. The corresponding feature flag will be removed and the feature will be tied to the appropriate `/langversion` option.  
+
+It's important to note there are **no guarantees** about a feature being a part of the language until it has reached the 'Acceptance' state **and** shipped in an official build. Features in any other state--no matter how complete--can be pulled at the last minute if the Language Team deems it's necessary to do so.  
+
+Pulling a feature after the prototype phase completes is not a decision the Langauge Team takes lightly; however, in the past, unforeseen circumstances have caused us to do so. The best example of this was the decision to pull parameterless struct constructors very late in the C# 6.0 timeframe.
```
Comment (bkoelman) [inline docs/contributing/Developing a Language Feature.md]: Language Team

**Comment by davkean** [comment]
Mac is back up.
@dotnet-bot test prtest/mac/dbg/unit32 please

**Pull request merged**